crawl-ref/source/ghost.cc

```diff
@@ -172,6 +167,194 @@ static int _panlord_random_elec_resist_level()
                                   1, 3);
 }
 
+/**
+ * Generate a random attack_type for a pandemonium_lord. Since this is purely
+ * flavour, special attack types are rare.
+ */
+static attack_type _pan_lord_random_attack_type()
+{
+    attack_type attack = AT_HIT;
+    if (one_chance_in(4))
+    {
+        do {
+            attack = static_cast<attack_type>(random_range(AT_FIRST_ATTACK, AT_LAST_REAL_ATTACK));
+        } while (attack == AT_HIT || !is_plain_attack_type(attack));
+    }
+    return attack;
+}
+
+/**
+ * Set a random attack type for a pandemonium lord's melee attacks. Some of
+ * these are branded attacks, some are custom attack flavours, some are matching
+ * attack types & flavours. (Pan lord attack type is randomised, but can be
+ * overridden here if required.)
+ */
+void ghost_demon::set_pan_lord_special_attack()
```

Comment:
If there's a better way to implement this function I'm all ears! The current implementation feels "safe" but is incredibly verbose.